streams/shared/src/main/scala/zio/stream/internal/ChannelExecutor.scala

```diff
@@ -210,9 +210,16 @@ private[stream] final class ChannelExecutor[Env, InErr, InElem, InDone, OutErr,
             case ZChannel.ConcatAll(combineSubK, combineSubKAndInner, value, k) =>
               val innerExecuteLastClose =
                 (f: URIO[Env, Any]) =>
-                  ZIO.succeed {
-                    closeLastSubstream = if (closeLastSubstream eq null) f else closeLastSubstream *> f
```

Comment:
Part of the "leak" was being caused by this. On each `ZStream#flatMap` we were adding a new `*> ZIO.uninterruptible(Exit.unit)`